keps/sig-storage/3333-reconcile-default-storage-class/README.md

```diff
@@ -0,0 +1,656 @@
+    * PVC update from  `pvc.spec.storageClassName=nil` to
+      `pvc.spec.storageClassName=<storage class name>` is now forbidden, and it
+      must be allowed for this KEP to work.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+Kubernetes already has a good coverage of PV/PVC binding tests in
+[test/e2e/storage/persistent_volumes.go](https://github.com/kubernetes/kubernetes/blob/ccfac6d3200f63656575819e7b5976c12c3019a6/test/e2e/storage/persistent_volumes.go)
+
+There are no e2e tests that cover behavior of the default SC presence / absence,
+they will be added only with the new behavior.
+
+##### Unit tests
+
+All changes should be only in the packages below, which have enough unit test
+coverage, we will add only unit tests for the new or changed code.
+
+- `pkg/controller/volume/persistentvolume`: 2022-06-03 - 79%
+- `pkg/apis/core/validation/`: 2022-06-03 - 82%
+
+##### Integration tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- test/integration/volume/persistent_volumes_test.go: Has few cases for PV/PVC
+  binding.
+
+We plan to extend these test to include default SC and how it will be applied to
+existing PVCs. We use integration tests almost like e2e tests of the new
+behavior to be sure that change of the default SC won't affect other tests
+running in parallel.
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+There are no tests that would check how the default SC, because it would need to
+create/delete default SCs, which could affect other tests running in parallel
+that may expect that a default SC exists (typically StatefulSet e2e tests).
+
+We plan to add a few `[Disruptive]` `[Serial]` tests with the new behavior as
+"smoke" tests of the new behavior, still, most of the tests will be integration
+ones.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag.
+- Initial integration tests completed and enabled.
+
+#### Beta
+
+- Implement and enable e2e tests, visible in Testgrid and linked in KEP.
+
+#### GA
+
+- No users complaining about the new behavior.
+- Scalability tests with existing framework (Kubemark?, TBD).
```

Comment:
Kubemark doesn't really support PVs at this point.

I think we mostly should be able to reuse the existing tests that we run against real cluster.
The existing test is creating a bunch of PDs - if we would just delete the default SC at the beginning of the test and recreate it somewhere later and check if PVC will get bound at this point - we should be fine.

Those should be relatively small change to do in the current test to test that.

[Also, given features are enabled by default in Beta, this should be a Beta criteria.]

Reply:
I found some pod-with-PV tests in [kubernetes/perf-tests](https://github.com/kubernetes/perf-tests/tree/81c96c34e3c1f11c5fe91744ef7ce7bf44c2fe5c/clusterloader2/testing/experimental/storage/pod-startup/volume-types/persistentvolume).  And yes, we could create a default SC somewhere in the middle of the test and see how it works.

Reply:
Done. Replaced Kubemark with k/perf-tests for scalability testing.